Addition of rule for cases better represented with assertSameSize() method

o0h opened the pull request:
This Pull Request enhances the AssertSameWithCountRule in PHPUnit to handle additional use cases. The new rule now checks when both sides of the assertSame are counts, and suggests assertSameSize instead when necessary.

Here is some detail on what has changed:
* AssertSameWithCountRule now includes a check for cases where expected value of assertCount is the direct result of count function call.
* In these cases, a special assertion function assertSameSize is suggested to compare the sizes of collections rather than using the original assertSame function.

The code changes involve adjustments in method call structures, evaluation of counts and introduction of cases for assertSameSize.

o0h committed:
Update AssertSameWithCountRule for assertSameSize

Expanded the AssertSameWithCountRule in PHPUnit rules to handle additional scenarios. Now the rule checks if both sides of the assertSame are counts, and suggests assertSameSize instead when necessary. The modification enhances the rule's applicability and accuracy.

ondrejmirtes reviewed (changes requested):
Hi,
this should only be reported with bleeding edge. Use `private bool $bleedingEdge` in the rule constructor and add the new behaviour only when this is true. Here's an example how it should look in config: https://github.com/phpstan/phpstan-doctrine/blob/0d967379b8ea58c4f4499b17eeb9f29718350497/rules.neon#L57

o0h commented:
Hi, @ondrejmirtes!

Thanks for the quick feedback! 
Sorry that it's not passing the CI checks just yet. I'll sort out the issue and let you know once it's ready for review.

I'll be sure to use the bleeding edge flag as you suggested. Just hang tight a little longer!

Best regards,

o0h committed:
Update README

o0h commented:
I have completed the necessary corrections that were of concern.
I would appreciate it if you could take a look and check it.
Thank you 😃

o0h committed:
Update link and fix typo in README.md

The link in the README file's "bleeding edge" reference has been updated to accurately direct users to the relevant blog post. Also corrected the typo by capitalizing the initial letter of the sentence.

ondrejmirtes commented on src/Rules/PHPUnit/AssertSameWithCountRule.php:
```diff
@@ -37,36 +46,89 @@ public function processNode(Node $node, Scope $scope): array
 		}
 
-		if (
-			$right instanceof Node\Expr\MethodCall
```
TBH this diff is very messy. I can't tell if the no-bleedingEdge behaviour is preserved or not. Can you split the changes in multiple commits? First extract logic into `isCount...` methods without changing the behaviour, and then introduce `bleedingEdge-on` behaviour that will be a new separate `if` statement. Can you please do that? Thank you.
